Add rate limiter processor

pavan-traceable opened the pull request:
## Description
Add rate limiter processor which helps in rate limiting spans which uses envoy based rate limit service. 


- **Rate Limit Support**: Add rate limit processor which rate limits the requests based on span count and it needs rate limit service deployed using https://github.com/envoyproxy/ratelimit. 
- There are tenant level limits and cluster level limits. For cluster level there is soft limit and hard limit. All spans will be dropped if cluster hard limit reached. If soft limit reached, requests will be dropped from tenants where tenant level limit reached.


### Testing
ratelimitprocessor_test has tests which tests following scenarios.
1. When rate limit service not available
2. When tenant header is missing.
3. When cluster level soft limit reached but hard limit not reached.
4. When cluster level hard limit reached. 



<!--
Include __important__ links regarding the implementation of this PR.
Rate limit processor requires https://github.com/envoyproxy/ratelimit
-->

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+		},
+	}
+	spanCount := uint32(traces.SpanCount())
+	response, err := p.rateLimitServiceClient.ShouldRateLimit(
```
Making an external call is going to slow down the HT Collector significantly. I'd prefer if this were done somewhere else or a network call inside the processor was avoided.

pavan-traceable replied:
This is already happening in istio-ingress. We have moved this to Hypertrace collector since we know the span count, which is not possible in Istio.

prodion23 replied:
@tim-mwangi thoughts about doing this async, similar to how we do ip-resolution? 

Downside would be we wouldn't rate limit immediately, but checking async would prevent slowdown.

tim-mwangi replied:
mmh i think the rate limiter service internal counters are also updated by making the call. @pavan-traceable does the rate-limiter consult any DBs? or is the data cached in memory.

pavan-traceable replied:
RateLimitService uses redis as backend store. https://github.com/envoyproxy/ratelimit this is the service which will be used.

tim-mwangi replied:
@pavan-traceable proceed with the impl but note we have to thoroughly test for performance and error scenarios when rate limit service is not responsive. Given that it's one request per traces batch, the impact of the latency may not be so bad.

pavan-traceable replied:
Sure @tim-mwangi

tim-mwangi reviewed (changes requested):
Not sure what the motivation for this change is and where the processor would be placed in the pipeline but introducing a network call synchronous to the spans processing will definitely cause performance issues. Also there are no metrics associated with the drops.

pavan-traceable commented:
> Not sure what the motivation for this change is and where the processor would be placed in the pipeline but introducing a network call synchronous to the spans processing will definitely cause performance issues. Also there are no metrics associated with the drops.

Metrics are unit tests for this, i am working on. I have put up a PR just to get initial comments.

shashank11p commented on processors/ratelimiter/factory.go:
```diff
@@ -0,0 +1,88 @@
+		params.Logger.Error("failed to connect to rate limit service ", zap.Error(err))
+		return nil, err
+	}
+	processor := &processor{
```
the tenantId header name is missing

pavan-traceable replied:
Added

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+	return traces, nil
+}
+
+func (p *processor) getTenantId(ctx context.Context) (string, error) {
```
Also the tenantId is also already extracted in the tenant id processor and should be one of the resource attributes in the resource spans: https://github.com/hypertrace/hypertrace-collector/blob/67cfc3a5d8e31593eb1bbfae2f78f7f5f580d2c5/processors/tenantidprocessor/tenantidprocessor.go#L75

pavan-traceable replied:
Just avoiding dependency on tenantId processor before this. Since tenantId extraction is straight forward

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+
+const (
+	TenantID       = "tenant_id"
+	GlobalTenantID = "global_tenant_id"
```
what's global tenant id?

pavan-traceable replied:
Some tenantId which represents cluster level count. Using this cluster level limits can be tracked.

tim-mwangi replied:
ok.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+const (
+	TenantID       = "tenant_id"
```
Is this default the same as with the tenant id processor? If we will extract the tenant id in the context here we need to have a way to configure the header key value to read from in the processor config.

pavan-traceable replied:
updated in factory.go

tim-mwangi replied:
ok

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+		&pb.RateLimitRequest{
+			Domain:      p.domain,
+			Descriptors: desc,
+			HitsAddend:  spanCount,
```
HitsAdded?

pavan-traceable replied:
It's count which needs to be incremented for descriptors.

tim-mwangi replied:
Oh I was referring to the typo.

tim-mwangi replied:
Ignore this. Just realized it's in the api.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+	if descriptorStatuses != nil {
+		if descriptorStatuses[1].GetCode() == pb.RateLimitResponse_OVER_LIMIT {
+			// If global rate limit exceeded drop all events
+			return traces, fmt.Errorf("dropping spans for tenant %s due to global rate limit exceeded, of spancount: %d", tenantId, traces.SpanCount())
```
`return traces,...` will not drop. You need `return nil, nil` to drop and not return an error so that it can be a successful return. Advise to keep metrics for this and also log that you have dropped - rate limited logging eg. once a tenant, a minute or something.

pavan-traceable replied:
Added logging and metrics as well. Updated this method to ConsumeTraces.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+)
+
+// ProcessTraces implements processorhelper.ProcessTracesFunc
+func (p *processor) ProcessTraces(ctx context.Context, traces ptrace.Traces) (ptrace.Traces, error) {
```
You might want to implement ConsumeTraces actually. That way you can fail to invoke `p.nextConsumer.ConsumeTraces(ctx, td)` if you want to drop

pavan-traceable replied:
Updated

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,95 @@
+	}
+	spanCount := uint32(traces.SpanCount())
+	response, err := p.rateLimitServiceClient.ShouldRateLimit(
+		ctx,
```
We should set a very short but reasonable `WithTimeout` for this so that we don't have the default 10s or wait on the rate limit service to fail to respond.

pavan-traceable replied:
Made timeout configurable and make default 1 second.

tim-mwangi commented on processors/ratelimiter/config.go:
```diff
@@ -0,0 +1,27 @@
+	// DomainSoftRateLimitThreshold represents the threshold where soft limit window started.
+	DomainSoftRateLimitThreshold uint32 `mapstructure:"domain_soft_limit_threshold"`
+	// TenantIDHeaderName defines tenant HTTP header name. Default x-tenant-id.
+	TenantIDHeaderName string `mapstructure:"header_name"`
```
Rename map structure field name to "tenant_id_header_name"

pavan-traceable replied:
done

tim-mwangi commented on go.sum:
```diff
@@ -786,6 +786,8 @@ github.com/stretchr/objx v0.1.0/go.mod h1:HFkY916IF+rwdDfMAkV7OtwuqBVzrE8GR6GFx+
 github.com/stretchr/objx v0.1.1/go.mod h1:HFkY916IF+rwdDfMAkV7OtwuqBVzrE8GR6GFx+wExME=
 github.com/stretchr/objx v0.4.0 h1:M2gUjqZET1qApGOWNSnZ49BAIMX4F/1plDv3+l31EJ4=
 github.com/stretchr/objx v0.4.0/go.mod h1:YvHI0jy2hoMjB+UWwv71VJQ9isScKT/TqJzVSSt89Yw=
+github.com/stretchr/objx v0.5.0 h1:1zr/of2m5FGMsad5YfcqgdqdWrIhu+EBEJRhR1U7z/c=
```
Can you run a `go mod tidy` to see if some of these old dependencies clear up?

pavan-traceable replied:
done

tim-mwangi commented on processors/ratelimiter/factory.go:
```diff
@@ -0,0 +1,87 @@
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```
I don't think you want to close the connection at the end of this function.

pavan-traceable replied:
Any idea how to handle this cleanly ?

tim-mwangi replied:
https://github.com/hypertrace/hypertrace-collector/pull/102#discussion_r1090733706

pavan-traceable replied:
Fixed it.

tim-mwangi commented on processors/ratelimiter/factory.go:
```diff
@@ -0,0 +1,87 @@
+	params.Logger.Info("connecting to rate limit service %s " + dialString)
+	conn, err = grpc.DialContext(ctxWithTimeout, dialString, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
```
log error

pavan-traceable replied:
done

tim-mwangi commented on processors/ratelimiter/factory.go:
```diff
@@ -0,0 +1,87 @@
+func getRateLimitServiceClient(ctx context.Context, serviceHost string, servicePort uint16,
+	timeoutMillis uint32, params component.ProcessorCreateSettings) (pb.RateLimitServiceClient, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, time.Millisecond*time.Duration(timeoutMillis))
+	defer cancel()
```
You also do not want to cancel this context at the end of the function.

pavan-traceable replied:
Thanks, got this doubt. Any idea when and how to cancel the context ?

tim-mwangi replied:
You can add a WithShutdown option when calling NewTracesProcessor https://github.com/open-telemetry/opentelemetry-collector/blob/5f1733ad50c7c7a2c7c320ea26cf1e5cd718013c/processor/processorhelper/processor.go#L46

pavan-traceable replied:
Updated.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,120 @@
+// ProcessTraces implements processorhelper.ProcessTracesFunc
+func (p *processor) ConsumeTraces(ctx context.Context, traces ptrace.Traces) error {
+	tenantId, err := p.getTenantId(ctx)
+	if err != nil {
```
log error

pavan-traceable replied:
done

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,120 @@
+		if descriptorStatuses[1].GetCode() == pb.RateLimitResponse_OVER_LIMIT {
+			// If global rate limit exceeded drop all events
+			p.logger.Warn(fmt.Sprintf("dropping spans for tenant %s due to global rate limit exceeded, of spancount: %d", tenantId, traces.SpanCount()))
+			stats.Record(ctx, droppedSpanCount.M(int64(spanCount)))
```
We need to associate the metric with the tenant id. See how the tenant id processor does it.

tim-mwangi replied:
Also can we distinguish this global rate limit drop from the soft rate limit one? by using 2 different metrics or labels.

pavan-traceable replied:
Updated, please check.

tim-mwangi commented:
You will want to register new metrics at https://github.com/hypertrace/hypertrace-collector/blob/67cfc3a5d8e31593eb1bbfae2f78f7f5f580d2c5/cmd/collector/main.go#L117

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func (p *processor) Shutdown(_ context.Context) error {
+	return nil
```
I think you can also put in the connection close and context cancelling here if you don't want to use options. Which means you have to track those objects in your processor.

pavan-traceable replied:
Done

tim-mwangi commented on processors/ratelimiter/config.go:
```diff
@@ -0,0 +1,27 @@
+	config.ProcessorSettings `mapstructure:"-"`
+
+	// RateLimitServiceHost defines host where rate limiter service is running default "localhost".
+	RateLimitServiceHost string `mapstructure:"service_host"`
```
No need for the `RateLimit` prefix for these config fields.

pavan-traceable replied:
Updated accordingly

tim-mwangi commented on processors/ratelimiter/config.go:
```diff
@@ -0,0 +1,27 @@
+	// TenantIDHeaderName defines tenant HTTP header name. Default x-tenant-id.
+	TenantIDHeaderName string `mapstructure:"tenant_id_header_name"`
+	// Timeout in millis for grpc call.
+	RateLimitServiceTimeoutMillis uint32 `mapstructure:"timeout_millis"`
```
Try to match the field name in the code to the JSON equivalent. So this should be "TimeoutMillis" or the config field name should change to "rate_limit_service_timeout_millis"

pavan-traceable replied:
Updated.

pavan-traceable commented:
@tim-mwangi  Can you please review now ? Addressed all the comments.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,156 @@
+
+var (
+	tagTenantID               = tag.MustNewKey("tenant-id")
+	droppedSpanCountSoftLimit = stats.Int64("tenant_id_dropped_span_count_soft_limit", "Number of spans dropped because of rate limiting per tenant due to soft limit", stats.UnitDimensionless)
```
fix this

pavan-traceable replied:
done

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,156 @@
+	droppedSpanCount          = stats.Int64("tenant_id_dropped_span_count", "Number of spans dropped per tenant due to global rate limiting", stats.UnitDimensionless)
+)
+
+func MetricViews() []*view.View {
```
Put this metrics logic in a separate go file. Called metrics.go or something like that.

pavan-traceable replied:
done

tim-mwangi commented on processors/ratelimiter/factory.go:
```diff
@@ -0,0 +1,90 @@
+	nextConsumer consumer.Traces,
+) (component.TracesProcessor, error) {
+	pCfg := cfg.(*Config)
+	rateLimitServiceClient, rateLimitServiceClientConn, cancelFunc, err := getRateLimitServiceClient(ctx, pCfg.ServiceHost, pCfg.ServicePort, pCfg.TimeoutMillis, params)
```
If you have declared default config values, are they merged into the incoming config? I don't think so but I could be wrong.

pavan-traceable replied:
You are right, not able to find where config merge happening, can you share reference for this ?

tim-mwangi replied:
I think you have to check for empty value manually and set them yourself.

pavan-traceable replied:
Not sure how it is happening but config override is happening  if it is in config.yaml. I have tested locally.

tim-mwangi replied:
Not config override. If you do not have timeoutmillis set, what's the config value passed on to the processor? Is it the default value of 1000ms?

pavan-traceable replied:
Yes.

tim-mwangi commented:
> ## Description
> Add rate limiter processor which helps in rate limiting spans which uses envoy based rate limit service.
> 
> ### Testing
> Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally.
> 
> ### Checklist:
> * [ ]  My changes generate no new warnings
> * [ ]  I have added tests that prove my fix is effective or that my feature works
> * [ ]  Any dependent changes have been merged and published in downstream modules
> 
> ### Documentation
> Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

Clean this up and answer these questions.

tim-mwangi commented:
No reference. I think you have to manually do it if the values in the
incoming config are not set.

On Mon, Jan 30, 2023, 20:43 pavan-traceable ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In processors/ratelimiter/factory.go
> <https://github.com/hypertrace/hypertrace-collector/pull/102#discussion_r1091444289>
> :
>
> > +		ServicePort:                  defaultServicePort,
> +		Domain:                       defaultDomain,
> +		DomainSoftRateLimitThreshold: defaultDomainSoftLimitThreshold,
> +		TenantIDHeaderName:           defaultHeaderName,
> +		TimeoutMillis:                defaultTimeoutMillis,
> +	}
> +}
> +
> +func createTraceProcessor(
> +	ctx context.Context,
> +	params component.ProcessorCreateSettings,
> +	cfg config.Processor,
> +	nextConsumer consumer.Traces,
> +) (component.TracesProcessor, error) {
> +	pCfg := cfg.(*Config)
> +	rateLimitServiceClient, rateLimitServiceClientConn, cancelFunc, err := getRateLimitServiceClient(ctx, pCfg.ServiceHost, pCfg.ServicePort, pCfg.TimeoutMillis, params)
>
> You are right, not able to find where config merge happening, can you
> share reference for this ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hypertrace/hypertrace-collector/pull/102#discussion_r1091444289>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKVWBD5MAPWONYE5RAPCIRDWVCKATANCNFSM6AAAAAAUFAERYU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

pavan-traceable commented:
> > ## Description
> > Add rate limiter processor which helps in rate limiting spans which uses envoy based rate limit service.
> > ### Testing
> > Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally.
> > ### Checklist:
> > 
> > * [ ]   My changes generate no new warnings
> > * [ ]   I have added tests that prove my fix is effective or that my feature works
> > * [ ]   Any dependent changes have been merged and published in downstream modules
> > 
> > ### Documentation
> > Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.
> 
> Clean this up and answer these questions.

Updated

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+func (p *processor) Shutdown(_ context.Context) error {
+	p.rateLimitServiceClientConn.Close()
```
These close and cancel functions do not return errors? If they do we should not ignore them. We should log them.

pavan-traceable replied:
Fixed it.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,125 @@
+	GlobalTenantID = "global_tenant_id"
+)
+
+// ProcessTraces implements processorhelper.ProcessTracesFunc
```
Not true anymore :)

pavan-traceable replied:
my bad, updated.

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,125 @@
+	descriptorStatuses := response.Statuses
+	ctx, _ = tag.New(ctx,
+		tag.Insert(tagTenantID, tenantId))
+	if len(descriptorStatuses) != 0 {
```
If we are fetching `descriptorStatuses[1]` we should check that the length is at least 2.

pavan-traceable replied:
done

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,125 @@
+	if len(descriptorStatuses) != 0 {
+		if descriptorStatuses[1].GetCode() == pb.RateLimitResponse_OVER_LIMIT {
+			// If global rate limit exceeded drop all events
+			p.logger.Warn(fmt.Sprintf("dropping spans for tenant %s due to global rate limit exceeded, of spancount: %d", tenantId, traces.SpanCount()))
```
Use `spanCount` instead of `traces.SpanCount()`

pavan-traceable replied:
fixed

tim-mwangi commented on processors/ratelimiter/ratelimiterprocessor.go:
```diff
@@ -0,0 +1,125 @@
+			return nil
+		} else if descriptorStatuses[1].LimitRemaining < p.domainSoftLimitThreshold && descriptorStatuses[0].GetCode() == pb.RateLimitResponse_OVER_LIMIT {
+			// If soft limit reached then drop only spans from tenants where tenant specific limit exceeded.
+			p.logger.Warn(fmt.Sprintf("dropping spans for tenant %s due to soft limit reached, of spancount: %d", tenantId, traces.SpanCount()))
```
ditto

pavan-traceable replied:
fixed.

pavan-traceable commented:
@tim-mwangi  Can you please review again ?
